src/librustc_mir/hair/pattern/_match.rs

```diff
@@ -447,43 +617,660 @@ impl<'tcx> Constructor<'tcx> {
+                ty::Slice(ty) | ty::Array(ty, _) => (0..prefix + suffix).map(|_| ty).collect(),
+                _ => bug!("bad slice pattern {:?} {:?}", self, ty),
+            },
+            ConstantValue(_) | MissingConstructors(_) | ConstantRange(..) | Wildcard => vec![],
+        };
+
+        subpattern_types.into_iter().map(|ty| Pat { ty, span: DUMMY_SP, kind: box PatKind::Wild })
+    }
+
+    /// This computes the arity of a constructor. The arity of a constructor
+    /// is the number of its arguments.
+    ///
+    /// For instance, a tuple pattern `(_, 42, Some([]))` has arity 3, a struct pattern's arity is
+    /// the number of fields it contains, etc.
+    fn arity<'a>(&self, cx: &MatchCheckCtxt<'a, 'tcx>, ty: Ty<'tcx>) -> u64 {
+        debug!("Constructor::arity({:#?}, {:?})", self, ty);
+        match *self {
+            Single | Variant(_) => match ty.kind {
+                ty::Tuple(ref fs) => fs.len() as u64,
+                ty::Ref(..) => 1,
+                ty::Adt(adt, _) => {
+                    adt.variants[self.variant_index_for_adt(cx, adt)].fields.len() as u64
+                }
+                ty::Slice(..) | ty::Array(..) => bug!("bad slice pattern {:?} {:?}", self, ty),
+                _ => 0,
+            },
+            FixedLenSlice(length) => length,
+            VarLenSlice(prefix, suffix) => prefix + suffix,
+            ConstantValue(_) | ConstantRange(..) | Wildcard | MissingConstructors(_) => 0,
         }
     }
+
+    /// Apply a constructor to a list of patterns, yielding a new pattern. `pats`
+    /// must have as many elements as this constructor's arity.
+    ///
+    /// Examples:
+    /// self: Single
+    /// ty: tuple of 3 elements
+    /// pats: [10, 20, _]           => (10, 20, _)
+    ///
+    /// self: Option::Some
+    /// ty: Option<bool>
+    /// pats: [false]  => Some(false)
+    fn apply<'a>(
+        &self,
+        cx: &MatchCheckCtxt<'a, 'tcx>,
+        ty: Ty<'tcx>,
+        pats: impl IntoIterator<Item = Pat<'tcx>>,
+    ) -> SmallVec<[Pat<'tcx>; 1]> {
+        let mut pats = pats.into_iter();
+        let pat = match *self {
+            Single | Variant(_) => match ty.kind {
+                ty::Adt(..) | ty::Tuple(..) => {
+                    let pats = pats
+                        .enumerate()
+                        .map(|(i, p)| FieldPat { field: Field::new(i), pattern: p })
+                        .collect();
+
+                    if let ty::Adt(adt, substs) = ty.kind {
+                        if adt.is_enum() {
+                            PatKind::Variant {
+                                adt_def: adt,
+                                substs,
+                                variant_index: self.variant_index_for_adt(cx, adt),
+                                subpatterns: pats,
+                            }
+                        } else {
+                            PatKind::Leaf { subpatterns: pats }
```

Comment:
Could collapse this into the other `} else {` using a guard `if adt.is_enum() =>`.